core/packages/contracts/src/BeefyClient.sol

```diff
@@ -326,55 +311,20 @@ contract BeefyClient is Ownable {
         bytes32[] calldata leafProof,
         uint256 leafProofOrder
     ) public {
-        bytes32 commitmentHash = keccak256(encodeCommitment(commitment));
-        bytes32 ticketID = createTicketID(msg.sender, commitmentHash);
-        Ticket storage ticket = tickets[ticketID];
-
-        if (ticket.prevRandao == 0) {
-            revert PrevRandaoNotCaptured();
-        }
+        (bytes32 commitmentHash, bytes32 ticketID) = checkInternal(commitment, bitfield);
 
-        if (commitment.validatorSetID != nextValidatorSet.id) {
-            revert InvalidCommitment();
-        }
-
-        if (commitment.blockNumber <= latestBeefyBlock) {
-            revert StaleCommitment();
-        }
-
-        if (leaf.nextAuthoritySetID != nextValidatorSet.id + 1) {
-            revert InvalidMMRLeaf();
-        }
-
-        if (ticket.bitfieldHash != keccak256(abi.encodePacked(bitfield))) {
-            revert InvalidBitfield();
-        }
-
-        verifyCommitment(commitmentHash, bitfield, nextValidatorSet, ticket, proofs);
-
-        bool leafIsValid = MMRProof.verifyLeafProof(
-            commitment.payload.mmrRootHash, keccak256(encodeMMRLeaf(leaf)), leafProof, leafProofOrder
+        verifyInternalWithHandover(
+            commitment, bitfield, proofs, leaf, leafProof, leafProofOrder, commitmentHash, ticketID
         );
-        if (!leafIsValid) {
-            revert InvalidMMRLeafProof();
-        }
 
-        currentValidatorSet = nextValidatorSet;
-        nextValidatorSet.id = leaf.nextAuthoritySetID;
-        nextValidatorSet.length = leaf.nextAuthoritySetLen;
-        nextValidatorSet.root = leaf.nextAuthoritySetRoot;
-
-        latestMMRRoot = commitment.payload.mmrRootHash;
-        latestBeefyBlock = commitment.blockNumber;
-
-        emit NewMMRRoot(commitment.payload.mmrRootHash, commitment.blockNumber);
-        delete tickets[ticketID];
+        applyInternalWithHandover(commitment, leaf, ticketID);
```

Comment:
I think this too much factoring. So while we've now eliminated duplication, we've also made the code flow harder to understand, because the logic has been split into too many little functions. It's a tricky balance to get correct 😅

However, the `checkInternal` function you created makes sense. I like it. That's a good level of refactoring.

So in conclusion, lets revert back to the previous versions of `submitFinal*`, and then convert them to use `checkInternal`.

Reply:
Yes, seems a bit over refactoring.